client/plugins/store/persistedStatePlugin.ts

```diff
@@ -0,0 +1,68 @@
+import { Plugin } from '@nuxt/types';
+import createPersistedState from 'vuex-persistedstate';
+import Cookies from 'js-cookie';
+import cookie from 'cookie';
+import destr from 'destr';
+
+const SPAStores = ['deviceProfile', 'clientSettings', 'i18n'];
+const SSRStores = ['user', 'servers', 'auth'];
+
+const persistedStatePlugin: Plugin = ({ store, req, res }) => {
+  /**
+   * Checks if client was built with SSR or as an SPA application.
+   *
+   * process.server or process.browser lets us distinguish between both environments when running SSR,
+   * but the check client-side just lets us know that we're in the browser environment, without further
+   * information on whether browser there's SSR behind the scenes.
+   *
+   * A hardcoded variable provides us that information, which helps us to manage the Vuex persisted storage properly
+   * by saving all the store's information in localStorage when the client is fully SPA and combining cookies and localStorage
+   * when using SSR, as SSR rendering needs the server to have access to the authentication data so the redirection
+   * middleware acts accordingly.
+   *
+   * Cookies pollutes a lo the headers of the clients and are problematic privacy-wise, so it's a good idea to avoid using them
```

Comment:
The thing is that they're not needed at all when client is built without SSR, so we're sending a bunch of data in each request done (yes, the cookie header gets long when you add multipe servers), even being unnecessary in a client-side scenario, so we're wasting resources unnecessarily when the client is built without SSR.

With this approach we keep the same thing we've been doing now for SSR (with cookies for auth stuff) without affecting SPA mode (being localStorage-only now), as the cookies were implemented on the SSR switch because server-side needs to know the auth details to run the redirection middleware

There are also some stuff I don't agree with the article:

* Having cookies available to subdomains can introduce inconsistencies between what we have in the store and the user that logged in. For instance, changing subdomains will still keep the user logged in but displaypreferences, homesections (once they're customizable) will be different because the mismatch between both. Having everything in Vuex ensures a consistent state
* Expiration should be handled server-side and not client-side in my opinion. If we want to have expirations, that should be controlled server-side, and not have our own limits, this can be confusing for users.
* We don't need to manage anything from localStorage as that's already handled by Nuxt Auth using Vuex. Nuxt Auth already invalidates everything as soon as it detects the session was revoked server side.


Of course all this stuff doesn't apply to clients built in SSR which will require cookies, that won't change. All I'm talking about is SPA mode of course.

Reply:
> Having cookies available to subdomains can introduce inconsistencies between what we have in the store and the user that logged in.

You are misunderstanding how the cross-subdomain thing works, here. Per the [spec](https://datatracker.ietf.org/doc/html/rfc6265#section-5.1.3):

```
5.1.3.  Domain Matching

   A string domain-matches a given domain string if at least one of the
   following conditions hold:

   o  The domain string and the string are identical.  (Note that both
      the domain string and the string will have been canonicalized to
      lower case at this point.)

   o  All of the following conditions hold:

      *  The domain string is a suffix of the string.

      *  The last character of the string that is not included in the
         domain string is a %x2E (".") character.

      *  The string is a host name (i.e., not an IP address).
```

So a cookie set for `vue.example.com` will only be readable by `vue.example.com`, while `.example.com` will be readable by any subdomain of `example.com`, and `example.com` is readable only by `example.com`.

> Expiration should be handled server-side and not client-side in my opinion.

I really disagree with this.

What happens if I login at a friend's house and forget to disconnect? They'll have access to the session pretty much forever.

The point of expiration is that the client can "timeout" if the cookie isn't used for a while. So if I don't use my session for a few days, it will actively ask me to re-login, which your localStorage way won't do, since it only expires if the session on the server does. There's a reason why, even with localStorage available for years, everybody still uses cookies to store sessions (Except Emby and jf-web).

> We don't need to manage anything from localStorage as that's already handled by Nuxt Auth using Vuex.

And we don't need to manage cookies either, except parsing them on the server (which you need to do anyway). I fail to see how this changes anything?

---

Point is:

* We should use cookies for auth stuff
* And use localStorage for non-auth stuff (Saved servers, for example)